File: experiments-local/experiment-18/runs/agent=omp_language=python_model=openrouter/anthropic/claude-sonnet-4.6_tooling=none/rep3/brazil_soccer_mcp/formatting.py
"""
================================================================================
brazil_soccer_mcp.formatting
================================================================================
Context:
    Renders query results from the knowledge graph into the human-readable text
    blocks shown in the specification's "Example answer format" sections. The
    MCP tools call these so an LLM receives clean, consistent prose rather than
    raw structures.

Each renderer takes already-computed data (the graph does the work) plus the
KnowledgeGraph for display-name lookups, and returns a string.
================================================================================
"""

from __future__ import annotations

from typing import List, Optional, Tuple

from .graph import KnowledgeGraph, TeamRecord
from .loaders import Match, Player


def _score(m: Match) -> str:
    h = "?" if m.home_goal is None else str(m.home_goal)
    a = "?" if m.away_goal is None else str(m.away_goal)
    return f"{h}-{a}"


def _round_label(m: Match) -> str:
    parts = [m.competition]
    if m.stage:
        parts.append(m.stage.title())
    elif m.round:
        parts.append(f"Round {m.round}")
    if m.season:
        parts.append(str(m.season))
    return " ".join(parts)


def format_match_line(m: Match) -> str:
    d = m.date.isoformat() if m.date else "unknown date"
    return f"- {d}: {m.home} {_score(m)} {m.away} ({_round_label(m)})"


def format_matches(
    graph: KnowledgeGraph,
    matches: List[Match],
    title: Optional[str] = None,
    show: int = 15,
    head_to_head: Optional[dict] = None,
) -> str:
    if not matches:
        return (title + "\n" if title else "") + "No matches found."
    lines: List[str] = []
    if title:
        lines.append(title)
    for m in matches[:show]:
        lines.append(format_match_line(m))
    if len(matches) > show:
        lines.append(f"- ... ({len(matches) - show} more matches in dataset)")
    if head_to_head:
        h = head_to_head
        lines.append("")
        lines.append(
            f"Head-to-head in dataset: {h['team_a']} {h['a_wins']} wins, "
            f"{h['team_b']} {h['b_wins']} wins, {h['draws']} draws "
            f"(goals {h['a_goals']}-{h['b_goals']})"
        )
    return "\n".join(lines)


def format_team_record(
    team: str, rec: TeamRecord, scope: str = ""
) -> str:
    header = f"{team} record{(' (' + scope + ')') if scope else ''}:"
    return "\n".join(
        [
            header,
            f"- Matches: {rec.matches}",
            f"- Wins: {rec.wins}, Draws: {rec.draws}, Losses: {rec.losses}",
            f"- Goals For: {rec.goals_for}, Goals Against: {rec.goals_against} "
            f"(diff {rec.goal_difference:+d})",
            f"- Points: {rec.points}",
            f"- Win rate: {rec.win_rate * 100:.1f}%",
        ]
    )


def format_head_to_head(graph: KnowledgeGraph, h: dict, show: int = 10) -> str:
    lines = [f"{h['team_a']} vs {h['team_b']} head-to-head:"]
    played = [m for m in h["meetings"] if m.home_goal is not None]
    for m in played[-show:]:
        lines.append(format_match_line(m))
    if len(played) > show:
        lines.append(f"- ... ({len(played) - show} earlier meetings)")
    lines.append("")
    lines.append(
        f"Total: {h['team_a']} {h['a_wins']}W, {h['team_b']} {h['b_wins']}W, "
        f"{h['draws']} draws. Goals: {h['a_goals']}-{h['b_goals']} "
        f"over {len(played)} decided matches."
    )
    return "\n".join(lines)


def format_standings(
    graph: KnowledgeGraph,
    competition: str,
    season: int,
    rows: List[Tuple[str, TeamRecord]],
    show: int = 20,
) -> str:
    if not rows:
        return f"No standings available for {competition} {season}."
    lines = [f"{season} {competition} Standings (calculated from matches):"]
    for i, (key, rec) in enumerate(rows[:show], start=1):
        tag = " - Champion" if i == 1 else ""
        lines.append(
            f"{i}. {graph.team_display(key)} - {rec.points} pts "
            f"({rec.wins}W, {rec.draws}D, {rec.losses}L)"
            f"{tag}"
        )
    return "\n".join(lines)


def format_players(
    players: List[Player], title: Optional[str] = None, show: int = 15
) -> str:
    if not players:
        return (title + "\n" if title else "") + "No players found."
    lines: List[str] = []
    if title:
        lines.append(title)
    for i, p in enumerate(players[:show], start=1):
        bits = [f"Overall: {p.overall}"] if p.overall is not None else []
        if p.position:
            bits.append(f"Position: {p.position}")
        if p.club:
            bits.append(f"Club: {p.club}")
        if p.nationality:
            bits.append(f"Nationality: {p.nationality}")
        lines.append(f"{i}. {p.name} - " + ", ".join(bits))
    if len(players) > show:
        lines.append(f"... ({len(players) - show} more)")
    return "\n".join(lines)


def format_player_detail(p: Player) -> str:
    return "\n".join(
        [
            f"{p.name}",
            f"- Age: {p.age}  Nationality: {p.nationality}",
            f"- Club: {p.club}  Position: {p.position}  "
            f"Jersey: {p.jersey or 'n/a'}",
            f"- Overall: {p.overall}  Potential: {p.potential}",
            f"- Height: {p.height}  Weight: {p.weight}",
        ]
    )


def format_aggregate(stats: dict, label: str) -> str:
    return "\n".join(
        [
            f"Aggregate statistics ({label}):",
            f"- Matches: {stats['matches']}",
            f"- Total goals: {stats['total_goals']}",
            f"- Average goals per match: {stats['avg_goals_per_match']}",
            f"- Home win rate: {stats['home_win_rate'] * 100:.1f}%",
            f"- Away win rate: {stats['away_win_rate'] * 100:.1f}%",
            f"- Draw rate: {stats['draw_rate'] * 100:.1f}%",
        ]
    )


def format_biggest_wins(matches: List[Match], label: str) -> str:
    if not matches:
        return f"No matches found for {label}."
    lines = [f"Biggest victories ({label}):"]
    for i, m in enumerate(matches, start=1):
        margin = abs((m.home_goal or 0) - (m.away_goal or 0))
        lines.append(
            f"{i}. {m.date.isoformat() if m.date else '?'}: "
            f"{m.home} {_score(m)} {m.away} "
            f"({m.competition}{(' ' + str(m.season)) if m.season else ''}, "
            f"margin {margin})"
        )
    return "\n".join(lines)


def format_best_record(
    graph: KnowledgeGraph,
    rows: List[Tuple[str, TeamRecord]],
    label: str,
    show: int = 10,
) -> str:
    if not rows:
        return f"No records found for {label}."
    lines = [f"Best records ({label}):"]
    for i, (key, rec) in enumerate(rows[:show], start=1):
        lines.append(
            f"{i}. {graph.team_display(key)} - win rate "
            f"{rec.win_rate * 100:.1f}% "
            f"({rec.wins}W, {rec.draws}D, {rec.losses}L in {rec.matches})"
        )
    return "\n".join(lines)

File: experiments-local/experiment-17/runs/agent=omp_language=csharp_model=openrouter/z-ai/glm-5.2_tooling=none/rep1-failed/Tools/Formatting.cs
// ============================================================================
// File: Tools/Formatting.cs
// ----------------------------------------------------------------------------
// Context: Shared text formatting helpers used by every MCP tool to render
// match lists, team records, and standings as the human-readable strings the
// MCP client returns to the LLM. Keeping the formatting here lets each tool
// stay a thin query-and-render layer.
// ============================================================================

using System.Globalization;
using System.Text;
using BrazilianSoccerMcp.Data;

namespace BrazilianSoccerMcp.Tools;

internal static class Formatting
{
    public static string DateOnly(DateTime? d) =>
        d?.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture) ?? "????-??-??";

    /// <summary>Render one match as "DATE: Home X-Y Away (Competition Round/Stage)".</summary>
    public static string FormatMatch(SoccerMatch m)
    {
        var sb = new StringBuilder();
        sb.Append("- ").Append(DateOnly(m.Date)).Append(": ");
        sb.Append(m.HomeTeamRaw).Append(' ')
          .Append(m.HomeGoals?.ToString(CultureInfo.InvariantCulture) ?? "?")
          .Append('-')
          .Append(m.AwayGoals?.ToString(CultureInfo.InvariantCulture) ?? "?")
          .Append(' ').Append(m.AwayTeamRaw);

        sb.Append(" (").Append(m.Competition);
        if (!string.IsNullOrEmpty(m.Round)) sb.Append(" Round ").Append(m.Round);
        else if (!string.IsNullOrEmpty(m.Stage)) sb.Append(" ").Append(m.Stage);
        sb.Append(')');

        return sb.ToString();
    }

    public static string FormatRecord(TeamRecord r, string label)
    {
        var winRate = r.WinRate.ToString("P1", CultureInfo.InvariantCulture);
        return $"""
               {label}:
               - Matches: {r.Matches}
               - Wins: {r.Wins}, Draws: {r.Draws}, Losses: {r.Losses}
               - Goals For: {r.GoalsFor}, Goals Against: {r.GoalsAgainst}
               - Points: {r.Points} | Win rate: {winRate}
               """;
    }

    public static string FormatStandings(List<TeamRecord> standings, string title)
    {
        var sb = new StringBuilder();
        sb.AppendLine(title);
        int pos = 1;
        foreach (var r in standings)
        {
            var champ = pos == 1 ? " - Champion" : "";
            var relegated = pos > standings.Count - 4 ? " - Relegated" : "";
            sb.AppendLine(CultureInfo.InvariantCulture,
                $"{pos}. {r.Team} - {r.Points} pts ({r.Wins}W, {r.Draws}D, {r.Losses}L, GF:{r.GoalsFor} GA:{r.GoalsAgainst}){champ}{relegated}");
            pos++;
        }
        return sb.ToString().TrimEnd();
    }
}
